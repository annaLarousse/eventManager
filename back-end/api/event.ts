import fs from 'fs';
import { Express, Request, Response } from 'express';
import { IEvent } from '../models/event';
import { IApiResponse } from '../models/apiResponse';

const dataFilePath: string = __dirname + '/../data/' + 'events.json';
module.exports = (app: Express) => {
  app.get('/events', (_req: Request, res: Response) => {
    res.contentType('application/json');
    const data: string = getEventsFromJSONFile();
    if (!data || !Object.keys(JSON.parse(data)).length)
      res.status(404).json('No events found!');
    else {
      res.status(200);
      res.end(data);
    }
  });

  app.post('/event', (req: Request, res: Response) => {
    res.contentType('application/json');
    let apiResponse: IApiResponse = {
      status: 201,
      message: 'Event added successfully!',
    };
    const newEvent: IEvent = {
      id: -1,
      name: req.body.name ? `${req.body.name}` : '',
      description: req.body.description ? `${req.body.description}` : '',
      startDate: req.body.startDate ? `${req.body.startDate}` : '',
      endDate: req.body.endDate ? `${req.body.endDate}` : '',
    };

    const errorMessage = doValidation(newEvent);
    if (errorMessage) {
      apiResponse = {
        status: 400,
        message: errorMessage,
      };
    }
    if (apiResponse.status !== 400) {
      res.contentType('text/plain');
      let dataString = getEventsFromJSONFile();
      const data = dataString ? JSON.parse(dataString) : {};
      const dataLength: number = Object.keys(data).length;
      const newId: number = dataLength + 1;

      newEvent.id = newId;
      (data[newEvent.id] as Object) = newEvent;

      dataString = JSON.stringify(data, undefined, 2);
      fs.writeFileSync(dataFilePath, dataString);
      apiResponse = {
        status: 201,
        message: 'Event added successfully!',
      };
    }
    res.writeHead(apiResponse.status);
    res.end(JSON.stringify(apiResponse));
  });
};

function getEventsFromJSONFile(): string {
  return fs.readFileSync(dataFilePath, 'utf8');
}

function doValidation(data: IEvent): string | null {
  if (!isValidDate(data.startDate)) {
    return 'Start date must be a valid date.';
  }
  if (!isValidDate(data.endDate)) {
    return 'End date must be a valid date.';
  }
  if (!data.name) {
    return 'Name must be set.';
  }
  if (data.name.length > 32) {
    return 'Event name cannot exceed 32 characters.';
  }
  if (data.endDate <= data.startDate) {
    return 'The end date have to be after the start date.';
  }
  return null;
}

function isValidDate(dateString: string): boolean {
  return !!Date.parse(dateString);
}
