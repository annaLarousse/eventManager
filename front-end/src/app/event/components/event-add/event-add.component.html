<h3>Add new Event</h3>

<form [formGroup]="form">
  <div class="event-add__form-container">
    <div class="form__event-name">
      <mat-form-field>
        <mat-label>Name</mat-label>
        <input
          matInput
          id="event-name"
          formControlName="name"
          placeholder="Ex. First day at Kumojin"
          required
          maxlength="32"
        />
      </mat-form-field>
    </div>

    <div class="form__event-start-date">
      <mat-form-field>
        <mat-label>Start date</mat-label>
        <input
          matInput
          id="event-start-date"
          formControlName="startDate"
          type="datetime-local"
          required
        />
      </mat-form-field>
    </div>

    <div class="form__event-end-date">
      <mat-form-field>
        <mat-label>End date</mat-label>
        <input
          matInput
          id="event-end-date"
          formControlName="endDate"
          type="datetime-local"
          required
        />
      </mat-form-field>
    </div>

    <div class="form__event-description">
      <mat-form-field>
        <mat-label>Description</mat-label>
        <textarea
          matInput
          id="event-description"
          formControlName="description"
        ></textarea>
      </mat-form-field>
    </div>
  </div>

  <div class="event-add__button-container">
    <button
      mat-raised-button
      color="primary"
      [disabled]="form?.invalid"
      (click)="onAddNewEvent()"
    >
      Save
    </button>
  </div>
</form>
