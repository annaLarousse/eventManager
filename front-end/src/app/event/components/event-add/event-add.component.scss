:host.event-add {
  display: flex;
  flex-direction: column;
  margin: 1rem;

  .event-add {
    &__form-container {
      display: grid;
      column-gap: 2rem;
      grid-template-areas:
        "name name . . ."
        "start-date end-date . . ."
        "description description . . .";

      .form {
        &__event-name {
          grid-area: name;
        }

        &__event-start-date {
          grid-area: start-date;
        }

        &__event-end-date {
          grid-area: end-date;
        }

        &__event-description {
          grid-area: description;
        }
      }

      .mat-form-field {
        width: 100%;
      }
    }
  }
}
