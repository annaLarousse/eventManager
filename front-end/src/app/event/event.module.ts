import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { MatTableModule } from '@angular/material/table';
import { EventListComponent } from './components/event-list/event-list.component';
import { EventPageComponent } from './components/event-page/event-page.component';

@NgModule({
  declarations: [EventPageComponent, EventListComponent],
  imports: [CommonModule, HttpClientModule, MatTableModule],
  exports: [EventPageComponent],
})
export class EventModule {}
