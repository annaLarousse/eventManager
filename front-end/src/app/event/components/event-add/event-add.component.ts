import { Component } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { take, tap } from 'rxjs';
import { IEvent } from '../../../../../../back-end/models/event';
import { EventService } from '../../services/event.service';

@Component({
  selector: 'app-event-add',
  templateUrl: './event-add.component.html',
  styleUrls: ['./event-add.component.scss'],
  host: { class: 'event-add' },
})
export class EventAddComponent {
  public form: FormGroup = new FormGroup({
    name: new FormControl(),
    description: new FormControl(),
    startDate: new FormControl(),
    endDate: new FormControl(),
  });

  constructor(private _eventService: EventService) {}

  public onAddNewEvent(): void {
    const data: IEvent = {
      ...this.form.value,
      startDate: new Date(this.form.value?.startDate).toISOString(),
      endDate: new Date(this.form.value?.endDate).toISOString(),
    };
    this._eventService
      .postEvent(data)
      .pipe(
        tap((notification) => {
          if (notification?.type === 'success') {
            this.form.reset();
          }
        }),
        take(1)
      )
      .subscribe();
  }
}
