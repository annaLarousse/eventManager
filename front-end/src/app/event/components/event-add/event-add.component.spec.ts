import { HttpClientTestingModule } from '@angular/common/http/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { of } from 'rxjs';
import { EventService } from '../../services/event.service';
import { EventAddComponent } from './event-add.component';

describe('EventAddComponent', () => {
  let component: EventAddComponent;
  let fixture: ComponentFixture<EventAddComponent>;

  let eventServiceMock = {
    postEvent: () =>
      of({ type: 'success', message: 'Event added successfully' }),
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [EventAddComponent],
      providers: [{ provide: EventService, useValue: eventServiceMock }],
      imports: [HttpClientTestingModule, ReactiveFormsModule],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(EventAddComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should display the add new event form', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('AddNewEvent()', () => {
    it('should call postEvent() of EventService', () => {
      const spyForm = jest.spyOn(component.form, 'reset');
      const spyPostEvent = jest.spyOn(eventServiceMock, 'postEvent');
      expect(spyPostEvent).not.toHaveBeenCalled();
      component.onAddNewEvent();
      expect(spyPostEvent).toHaveBeenCalled();
    });

    it('should reset the form after event added succeffully', () => {
      const spyForm = jest.spyOn(component.form, 'reset');
      component.onAddNewEvent();
      expect(spyForm).toHaveBeenCalled();
    });
  });
});
