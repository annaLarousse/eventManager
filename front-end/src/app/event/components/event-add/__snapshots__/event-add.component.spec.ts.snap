// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventAddComponent should display the add new event form 1`] = `
<app-event-add
  __ngContext__={[Function LRootView]}
  _eventService={[Function Object]}
  form={[Function FormGroup]}
>
  <h3>
    Add new Event
  </h3><form
    class="ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <div
      class="event-add__form-container"
    >
      <div
        class="form__event-name"
      >
        <mat-form-field>
          <mat-label>
            Name
          </mat-label>
          <input
            class="ng-untouched ng-pristine ng-invalid"
            formcontrolname="name"
            id="event-name"
            matinput=""
            maxlength="32"
            placeholder="Ex. First day at Kumojin"
            required=""
          />
        </mat-form-field>
      </div>
      <div
        class="form__event-start-date"
      >
        <mat-form-field>
          <mat-label>
            Start date
          </mat-label>
          <input
            class="ng-untouched ng-pristine ng-invalid"
            formcontrolname="startDate"
            id="event-start-date"
            matinput=""
            required=""
            type="datetime-local"
          />
        </mat-form-field>
      </div>
      <div
        class="form__event-end-date"
      >
        <mat-form-field>
          <mat-label>
            End date
          </mat-label>
          <input
            class="ng-untouched ng-pristine ng-invalid"
            formcontrolname="endDate"
            id="event-end-date"
            matinput=""
            required=""
            type="datetime-local"
          />
        </mat-form-field>
      </div>
      <div
        class="form__event-description"
      >
        <mat-form-field>
          <mat-label>
            Description
          </mat-label>
          <textarea
            class="ng-untouched ng-pristine ng-valid"
            formcontrolname="description"
            id="event-description"
            matinput=""
          />
        </mat-form-field>
      </div>
    </div>
    <div
      class="event-add__button-container"
    >
      <button
        color="primary"
        disabled=""
        mat-raised-button=""
      >
         Save 
      </button>
    </div>
  </form>
</app-event-add>
`;
